wolfcrypt/src/kdf.c

```diff
@@ -1470,4 +1472,280 @@ int wc_KDA_KDF_onestep(const byte* z, word32 zSz, const byte* fixedInfo,
 }
 #endif /* WC_KDF_NIST_SP_800_56C */
 
+#ifdef HAVE_CMAC_KDF
+/**
+ * \brief Performs the two-step cmac key derivation function (KDF) as
+ * specified in SP800-56C, section 5.1, in counter mode.
+ *
+ *            Z                               fixedInfo
+ *        ____|_________________________________|___________
+ *       |    |                                 |           |
+ *       |    ________________                ___________   |
+ * salt--|-> | Randomness     |              | Key       |  |
+ *       |   | Extract        | --Key_kdk--> | Expansion | -|-output-->
+ *       |    ----------------                -----------   |
+ *        --------------------------------------------------
+ *
+ * \param [in]  salt         The input keying material for cmac.
+ * \param [in]  salt_len     The size of the input keying material.
+ * \param [in]  z            The input shared secret (message to cmac).
+ * \param [in]  zSz          The size of the input shared secret.
+ * \param [in]  fixedInfo    The fixed information in the KDF.
+ * \param [in]  fixedInfoSz  The size of the fixed information.
+ * \param [out] output       The buffer to store the derived secret.
+ * \param [in]  outputSz     The desired size of the output secret.
+ * \param [in]  heap         The heap hint.
+ * \param [in]  devId        The device id.
+ *
+ * \return 0 if the KDF operation is successful.
+ * \return BAD_FUNC_ARG if the input parameters are invalid.
+ * \return negative error code if the KDF operation fails.
+ */
+int wc_KDA_KDF_twostep_cmac(const byte * salt, word32 salt_len,
+                            const byte* z, word32 zSz,
+                            const byte* fixedInfo, word32 fixedInfoSz,
+                            byte* output, word32 outputSz,
+                            void * heap, int devId)
+{
+    byte   Key_kdk[WC_AES_BLOCK_SIZE]; /* key derivation key*/
+    word32 kdk_len = sizeof(Key_kdk);
+    word32 tag_len = WC_AES_BLOCK_SIZE;
+    #ifdef WOLFSSL_SMALL_STACK
```

Comment:
Not required, but consider using the helper macros for small stack. You might need to rebase  to latest master to get these.

```
WC_DECLARE_VAR(cmac, Cmac, 1, 0);

WC_ALLOC_VAR_EX(cmac, Cmac, 1, heap, DYNAMIC_TYPE_CMAC, return MEMORY_E);

WC_FREE_VAR_EX(cert, heap, DYNAMIC_TYPE_CMAC);
```